libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -5698,705 +1043,235 @@ instance HasSpec fn a => Pretty (WithPrec (Specification fn a)) where
             , viaShow cant
             ]
 
-instance HasSpec fn a => Pretty (Specification fn a) where
+instance HasSpec a => Pretty (Specification a) where
   pretty = pretty . WithPrec 0
 
-instance HasSpec fn a => Show (Specification fn a) where
+instance HasSpec a => Show (Specification a) where
   showsPrec d = shows . pretty . WithPrec d
 
-instance Pretty (Var a) where
-  pretty = fromString . show
+-- ==========================================================
+-- Contexts
+-- ==========================================================
+
+data
+  Context
+    (s :: Symbol)
+    (t :: Symbol -> [Type] -> Type -> Type)
+    (dom :: [Type])
+    rng
+    hole
+  where
+  Context ::
+    (All HasSpec dom, HasSpec rng) =>
+    t s dom rng -> CList 'Pre dom hole y -> Context s t dom rng hole
 
-instance Pretty (Name fn) where
-  pretty (Name v) = pretty v
+data Ctx hole rng where
+  Ctx ::
+    (HasSpec hole, HasSpec rng, Logic s t dom rng) => Context s t dom rng hole -> Ctx hole rng
+  HOLE :: HasSpec a => Ctx a a
 
--- ======================================================================
--- Size and its 'generic' operations over Sized types.
--- ======================================================================
+data Mode = Pre | Post
 
--- type Size = Integer
+infixr 5 :<|
+infixr 5 :|>
+infixr 5 :<>
 
--- | Because Sizes should always be >= 0, We provide this alternate generator
---   that can be used to replace (genFromSpecT @Integer), to ensure this important property
-genFromSizeSpec :: (BaseUniverse fn, MonadGenError m) => Specification fn Integer -> GenT m Integer
-genFromSizeSpec integerSpec = genFromSpecT (integerSpec <> geqSpec 0)
+data CList (x :: Mode) (as :: [Type]) (hole :: Type) (y :: Type) where
+  End :: CList 'Post '[] h y
+  (:<|) :: Literal a => a -> CList 'Post as h y -> CList 'Post (a ': as) h y
+  (:<>) :: HasSpec y => Ctx x y -> (forall i j. CList 'Post as i j) -> CList 'Pre (y ': as) x y
+  (:|>) :: Literal a => a -> CList 'Pre as h y -> CList 'Pre (a ': as) h y
 
-data SizeFn (fn :: [Type] -> Type -> Type) as b where
-  SizeOf :: forall fn a. (Sized fn a, HasSpec fn a) => SizeFn fn '[a] Integer
+instance Show (CList mode as hole y) where
+  show clist = "(" ++ showCList clist
+    where
+      showCList :: forall m ds h z. CList m ds h z -> String
+      showCList End = "End)"
+      showCList (x :<| y) = show x ++ " :<| " ++ showCList y
+      showCList (hole :<> xs) = show hole ++ " :<> " ++ showCList xs
+      showCList (x :|> y) = show x ++ " :|> " ++ showCList y
```

Comment:
This code is quite a lot more complicated than the old implementation of `ListCtx`. How come?

Reply:
I thought it important that the CList be a single datatype, so that construction of a CList be simple to use
Hence the visual alignment of the constructors (:<>) the hole,  (:|>) the prefix to the left of the hole, and (:<|) sufffix to the right of the hole. It gets rid of the Value datatype, and we don't need to parameterize with Term and Value like we do if we use List. 

Maybe it is not simpler to define, but it is very simple to use when it is used in a propagate method
```
propagate (Context LookupW (k :|> HOLE :<> End)) spec = ...
```

Reply:
The old code would be something like `propagate LookupW (k :> CtxHOLE) spec = ...` that's in a way even simpler. But let's go with this for now and we can see if we can simplify it further in the next PR.

I like the idea of the constructors being evocative.

Reply:
We can figure this out in a follow on PR